packages/database/src/TupaiaDatabase.js

```diff
@@ -685,15 +695,31 @@ function addJoin(baseQuery, recordType, joinOptions) {
 
 function getColSelector(connection, inputColStr) {
   const jsonOperatorPattern = /->>?/g;
-  if (!jsonOperatorPattern.test(inputColStr)) return inputColStr;
+  if (jsonOperatorPattern.test(inputColStr)) {
+    const params = inputColStr.split(jsonOperatorPattern);
+    const allButFirst = params.slice(1);
+    const lastIndexOfLookupAsText = inputColStr.lastIndexOf('->>');
+    const lastIndexOfLookupAsJson = inputColStr.lastIndexOf('->');
+    const selector = lastIndexOfLookupAsText >= lastIndexOfLookupAsJson ? '#>>' : '#>';
+
+    // Turn `config->item->>colour` into `config #>> '{item,colour}'`
+    // For some reason, Knex fails when we try to convert it to `config->'item'->>'colour'`
+    return connection.raw(`?? ${selector} '{${allButFirst.map(() => '??').join(',')}}'`, params);
+  }
 
-  const params = inputColStr.split(jsonOperatorPattern);
-  const allButFirst = params.slice(1);
-  const lastIndexOfLookupAsText = inputColStr.lastIndexOf('->>');
-  const lastIndexOfLookupAsJson = inputColStr.lastIndexOf('->');
-  const selector = lastIndexOfLookupAsText >= lastIndexOfLookupAsJson ? '#>>' : '#>';
+  /**
+   * Special handling of COALESCE() - one of the {@link supportedFunctions} - to treat its arguments
+   * as identifiers individually rather than trying to treat ‘COALESCE(foo, bar)’ as a single
+   * identifier.
+   */
+  const coalescePattern = /^COALESCE\(.+\)$/;
+  if (coalescePattern.test(inputColStr)) {
+    const [, argsString] = inputColStr.match(/^COALESCE\((.+)\)$/);
+    const bindings = argsString.split(',').map(arg => arg.trim());
+    const identifiers = bindings.map(() => '??');
+
+    return connection.raw(`COALESCE(${identifiers})`, bindings);
+  }
 
-  // Turn `config->item->>colour` into `config #>> '{item,colour}'`
-  // For some reason, Knex fails when we try to convert it to `config->'item'->>'colour'`
-  return connection.raw(`?? ${selector} '{${allButFirst.map(() => '??').join(',')}}'`, params);
+  return inputColStr;
 }
```

Comment:
Two notes:

1. This implementation assumes we never have a column selector which **both** calls `COALESCE()` **and** uses JSON selectors. If that happens, this algorithm fails. Is this a concern?
2. Should I make it iterate through `supportedFunctions` to handle anything function in that array?

Reply:
1. Ah that's fine I think
2. Probably worth it I think! Would make this functionality a lot more future proof if we want to make more use of db functions. If you're happy to :pray:

Reply:
Hm, it occurs to me that, given some `supportedFunction`, it would be difficult to know whether to treat its arguments as `?` values or `??` identifiers

I _could_ just build in an assumption and always use `??`, but this would silently break any attempt to use a function that takes, say, a number. For example, if we have `ROUND(foo, 2)` Knex would complain that there’s no column called `'2'`

Unless there’s a clean way about this, I think I’d be inclined to manually handle each supported function?